AI Epoch - June substantial update 📊

veronikasamborska1994 opened the pull request.

lucasrodes commented on etl/steps/data/garden/artificial_intelligence/2024-06-03/epoch.py:
```diff
@@ -0,0 +1,161 @@
+    )
+
+    # Replace nans with Unspecified in each column to avoid issues when calculating sume of notable systems
+    columns = ["organization_categorization", "domain", "organization"]
+    for column in columns:
+        if tb[column].astype(str).str.contains("nan").any():
+            tb[column] = tb[column].replace("nan", "Not specified")
```
I think here you can get rid (i) of the for-loop and (ii) the if-clause:

```python
columns = ["organization_categorization", "domain", "organization"]
tb[columns] = tb[columns].replace("nan", "Not specified")

lucasrodes commented on etl/steps/data/garden/artificial_intelligence/2024-06-03/epoch.py:
```diff
@@ -0,0 +1,161 @@
+    # Convert publication date to a datetime objects
+    tb["publication_date"] = pd.to_datetime(tb["publication_date"])
+
+    # Calculate 'days_since_1949'
+    tb["days_since_1949"] = (tb["publication_date"] - pd.to_datetime("1949-01-01")).dt.days
+    tb = tb.dropna(subset=["days_since_1949"])
+
+    tb = tb.reset_index(drop=True)
+    tb["days_since_1949"] = tb["days_since_1949"].astype(int)
```
Could remove one line:

```python
tb["days_since_1949"] = (tb["publication_date"] - pd.to_datetime("1949-01-01")).dt.days.astype("Int64")
tb = tb.dropna(subset=["days_since_1949"])

tb = tb.reset_index(drop=True)

lucasrodes commented on etl/steps/data/garden/artificial_intelligence/2024-06-03/epoch.py:
```diff
@@ -0,0 +1,161 @@
+        ["training_compute__flop", "organization", "authors", "country__from_organization"],
+        axis=1,
+    )
+    tb = tb.set_index(["days_since_1949", "system"], verify_integrity=True).sort_index()
```
You can use `tb.format(["days_since_1949", "system"])` instead here.

lucasrodes commented on etl/steps/data/garden/artificial_intelligence/2024-06-03/epoch_aggregates_affiliation.py:
```diff
@@ -0,0 +1,75 @@
+
+from etl.helpers import PathFinder, create_dataset
+
+log = get_logger()
```
you don't need to import `get_logger` and use `log`. Instead, simply use object `paths.log`.

So you can do `paths.log.info` instead of `log.info`.

lucasrodes commented on etl/steps/data/garden/artificial_intelligence/2024-06-03/epoch_aggregates_affiliation.py:
```diff
@@ -0,0 +1,75 @@
+        tb_agg[col].metadata.origins = origins
+
+    # Set the short_name metadata of the table
+    tb_agg.metadata.short_name = SHORT_NAME
+
+    # Set the index to year and country
+    tb_agg = tb_agg.set_index(["year", "organization_categorization"], verify_integrity=True)
```
You can use `format` here:

```python
tb_agg = tb_agg.format(
  ["year", "organization_categorization"], 
  short_name=paths.short_name
)

lucasrodes commented on etl/steps/data/garden/artificial_intelligence/2024-06-03/epoch_aggregates_affiliation.py:
```diff
@@ -0,0 +1,75 @@
+
+# Get paths and naming conventions for current step.
+paths = PathFinder(__file__)
+SHORT_NAME = paths.short_name
```
don't need this

lucasrodes commented on etl/steps/data/garden/artificial_intelligence/2024-06-03/epoch.py:
```diff
@@ -0,0 +1,161 @@
+
+from etl.helpers import PathFinder, create_dataset
+
+log = get_logger()
```
you can use `paths.log` instead of `log`. That means, calling `paths.log.info` instead of `log.info`.

lucasrodes commented on etl/steps/data/garden/artificial_intelligence/2024-06-03/epoch_aggregates_countries.py:
```diff
@@ -0,0 +1,67 @@
+    tb_agg = geo.harmonize_countries(df=tb_agg, countries_file=paths.country_mapping_path)
+
+    # Set the index to year and country
+    tb_agg = tb_agg.set_index(["year", "country"], verify_integrity=True)
```
Suggestion: use `tb.format` function.

lucasrodes commented on etl/steps/data/grapher/artificial_intelligence/2024-06-03/epoch_aggregates_domain.py:
```diff
@@ -0,0 +1,42 @@
+    #
+    # Rename for plotting model domain as country in grapher
+    tb = tb.rename(columns={"domain": "country"})
+    tb = tb.set_index(["country", "year"], verify_integrity=True)
```
use `tb.format`

lucasrodes commented on etl/steps/data/grapher/artificial_intelligence/2024-06-03/epoch_aggregates_organizations.py:
```diff
@@ -15,16 +15,16 @@ def run(dest_dir: str) -> None:
     #
     # Load garden dataset.
-    ds_garden = cast(Dataset, paths.load_dependency("epoch_aggregates"))
+    ds_garden = cast(Dataset, paths.load_dependency("epoch_aggregates_organizations"))
```
Use `paths.load_dataset`

lucasrodes commented on etl/steps/data/meadow/artificial_intelligence/2024-06-03/epoch.py:
```diff
@@ -0,0 +1,82 @@
+    # Select the columns of interest
+    tb = tb[cols]
+    # Replace empty strings with NaN values
+    tb.replace("", np.nan, inplace=True)
```
In general, I would recommend either sticking always to `inplace` operations or non-inplace ones. My preference is not to use `inplace`. So:

```py
tb = tb.replace("", np.nan)

lucasrodes commented on etl/steps/data/meadow/artificial_intelligence/2024-06-03/epoch.py:
```diff
@@ -0,0 +1,82 @@
+    # Convert the training compute column to float
+    tb["Training compute (FLOP)"] = tb["Training compute (FLOP)"].astype(float)
+
+    # Replace the missing values in the system column with the organization column
+    tb.loc[tb["System"].isna(), "System"] = tb.loc[tb["System"].isna(), "Organization"]
+
+    # If organization column is NaN as well replace the missing values in the system column with the authors column
+    condition = tb["System"].isna() & tb["Organization"].isna()
+    tb.loc[condition, "System"] = tb.loc[condition, "Authors"]
```
Could simplify all of this with

```py
tb["System"] = tb["System"].fillna(tb["Organization"]).fillna(tb["Authors"])

lucasrodes commented on etl/steps/data/meadow/artificial_intelligence/2024-06-03/epoch.py:
```diff
@@ -0,0 +1,82 @@
+    #
+    # Create a new table and ensure all columns are snake-case.
+    #
+    tb = tb.underscore().set_index(["system", "publication_date"], verify_integrity=True).sort_index()
```
suggestion: use `tb.format`:

```py
tb = tb.format(["system", "publication_date"])

lucasrodes commented on snapshots/artificial_intelligence/2024-06-03/epoch.csv.dvc:
```diff
@@ -0,0 +1,36 @@
+    date_accessed: 2024-06-03
+    # License
+    license:
+      name: Public domain
```
please add a link to the license. If none available, please use a generic (https://creativecommons.org/public-domain/cc0/)

veronikasamborska1994 replied:
just took a look and it is actually CC BY 4.0 license so I'll change it! thank you!

lucasrodes reviewed (changes requested):
Good overall!

I've added some suggestions to improve the code quality. There are various tools that we have that you could use:

- `Table.format`: This function sets the specified index, verifies integrity and sorts the rows. Can also set the short name and more.
- `paths.log`: No need to import `get_logger` from `structlog`, you can use this custom method. So, instead of `log.info` use `paths.log.info`.
- `Table.rename_index_names`: You can rename index column names directly, without the need of resetting index.
- `paths.load_dataset`: Instead of using `paths.load_dependency`, we now have specific calls to load either a dataset or a snapshot.

I also see that there are some unit tests failing.

veronikasamborska1994 commented:
hey @

> Good overall!
> 
> I've added some suggestions to improve the code quality. There are various tools that we have that you could use:
> 
> * `Table.format`: This function sets the specified index, verifies integrity and sorts the rows. Can also set the short name and more.
> * `paths.log`: No need to import `get_logger` from `structlog`, you can use this custom method. So, instead of `log.info` use `paths.log.info`.
> * `Table.rename_index_names`: You can rename index column names directly, without the need of resetting index.
> * `paths.load_dataset`: Instead of using `paths.load_dependency`, we now have specific calls to load either a dataset or a snapshot.
> 
> I also see that there are some unit tests failing.

thanks a lot @lucasrodes! I think I've fixed everything now and tried to simplify the function you rightly suggested was messy and prone to doing something weird. also, sorry about all the `paths.log `, `paths.load_dependency` etc.

lucasrodes commented:
no problemo at all, happy to provide feedback! thanks for working on the changes :)!

veronikasamborska1994 commented:
thanks a lot @lucasrodes! btw something weird happened on this branch again - if you go to the `chart-diff` I am getting some charts from IHME, Global Burden of Disease (2024) dataset to review. I don't think I will use the `chart-diff` tool for this one as I've clearly done something weird before which is messing everything up but just so you know in case this is useful. Instead, I am going to try out the new workflow on this [pull request](https://github.com/owid/etl/pull/2773) which is a lot cleaner :)
